File: packages/sdk/src/resources/api-keys.ts
/**
 * API keys resource client. Maps to `/v1/api-keys`.
 *
 * The plaintext secret is returned exactly once on issuance.
 */
import type { ApiKey } from "@settlekit/common";
import type { HttpClient, RequestOptions } from "../http-client.js";

/** Input for {@link ApiKeysResource.issue}. */
export interface IssueApiKeyInput {
  organizationId: string;
  customerId: string;
  productId: string;
  entitlementId: string;
  scopes: string[];
  env?: "live" | "test";
}

/** Result of issuing an API key (plaintext shown once). */
export interface IssueApiKeyResult {
  apiKey: ApiKey;
  plaintext: string;
}

/** Input for {@link ApiKeysResource.verify}. */
export interface VerifyApiKeyInput {
  key: string;
  requiredScopes?: string[];
}

/** Result of verifying an API key. */
export interface VerifyApiKeyResult {
  valid: boolean;
  apiKey?: ApiKey;
}

/** Client for API key endpoints. */
export class ApiKeysResource {
  constructor(private readonly http: HttpClient) {}

  /** Issue a new scoped API key. Returns the one-time plaintext. */
  issue(input: IssueApiKeyInput, options?: RequestOptions): Promise<IssueApiKeyResult> {
    return this.http.post<IssueApiKeyResult>("/v1/api-keys", input, options);
  }

  /** Verify a presented key (and optional required scopes). */
  verify(input: VerifyApiKeyInput, options?: RequestOptions): Promise<VerifyApiKeyResult> {
    return this.http.post<VerifyApiKeyResult>("/v1/api-keys/verify", input, options);
  }

  /** Revoke a key by its plaintext value. */
  revoke(key: string, options?: RequestOptions): Promise<ApiKey> {
    return this.http.post<ApiKey>("/v1/api-keys/revoke", { key }, options);
  }
}

File: sdks/rust/src/resources/api_keys.rs
//! Scoped API-key resource (`/v1/api-keys`).

use reqwest::Method;
use serde::{Deserialize, Serialize};

use crate::{ApiKey, Client, Result};

/// Accessor for scoped API-key operations.
#[derive(Debug)]
pub struct ApiKeys<'a> {
    client: &'a Client,
}

/// Body for issuing an API key (`POST /v1/api-keys`).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct IssueApiKey {
    pub organization_id: String,
    pub customer_id: String,
    pub product_id: String,
    pub entitlement_id: String,
    pub scopes: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub env: Option<String>,
}

/// Issuance result: the stored key plus the one-time plaintext secret.
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IssuedApiKey {
    pub api_key: ApiKey,
    /// The plaintext secret, returned exactly once at issuance.
    pub plaintext: String,
}

/// Body for verifying an API key (`POST /v1/api-keys/verify`).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct VerifyApiKey {
    pub key: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub required_scopes: Option<Vec<String>>,
}

/// Outcome of verifying a presented key.
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VerifyApiKeyResult {
    pub valid: bool,
    #[serde(default)]
    pub api_key: Option<ApiKey>,
}

#[derive(Debug, Serialize)]
struct RevokeBody<'r> {
    key: &'r str,
}

impl<'a> ApiKeys<'a> {
    pub(crate) fn new(client: &'a Client) -> Self {
        ApiKeys { client }
    }

    /// Issue a new API key, returning the one-time plaintext.
    pub async fn issue(&self, body: &IssueApiKey) -> Result<IssuedApiKey> {
        self.client
            .request(Method::POST, "/v1/api-keys", Some(body))
            .await
    }

    /// Verify a presented key, optionally requiring scopes.
    pub async fn verify(&self, body: &VerifyApiKey) -> Result<VerifyApiKeyResult> {
        self.client
            .request(Method::POST, "/v1/api-keys/verify", Some(body))
            .await
    }

    /// Revoke a key by its plaintext.
    pub async fn revoke(&self, key: &str) -> Result<ApiKey> {
        let body = RevokeBody { key };
        self.client
            .request(Method::POST, "/v1/api-keys/revoke", Some(&body))
            .await
    }
}
